Optimize MVCCGet path.

petermattis opened the pull request:
Provide fast-paths for retrieving unmarshalling proto values that do not
copy the data into a Go slice. Avoid unnecessary allocations when
encoding mvcc keys.

benchmark                             old ns/op     new ns/op     delta
BenchmarkMVCCScan1Version1Row         11322         10721         -5.31%
BenchmarkMVCCScan1Version10Rows       36566         27190         -25.64%
BenchmarkMVCCScan1Version100Rows      276241        217066        -21.42%
BenchmarkMVCCScan1Version1000Rows     2205243       1834826       -16.80%
BenchmarkMVCCGet1Version              3739          3153          -15.67%

benchmark                             old MB/s     new MB/s     speedup
BenchmarkMVCCScan1Version1Row         90.44        95.51        1.06x
BenchmarkMVCCScan1Version10Rows       280.04       376.60       1.34x
BenchmarkMVCCScan1Version100Rows      370.69       471.74       1.27x
BenchmarkMVCCScan1Version1000Rows     464.35       558.09       1.20x
BenchmarkMVCCGet1Version              273.83       324.77       1.19x

petermattis committed:
Optimize MVCCGet path.

Provide fast-paths for retrieving unmarshalling proto values that do not
copy the data into a Go slice. Avoid unnecessary allocations when
encoding mvcc keys.

benchmark                             old ns/op     new ns/op     delta
BenchmarkMVCCScan1Version1Row         11322         10721         -5.31%
BenchmarkMVCCScan1Version10Rows       36566         27190         -25.64%
BenchmarkMVCCScan1Version100Rows      276241        217066        -21.42%
BenchmarkMVCCScan1Version1000Rows     2205243       1834826       -16.80%
BenchmarkMVCCGet1Version              3739          3153          -15.67%

benchmark                             old MB/s     new MB/s     speedup
BenchmarkMVCCScan1Version1Row         90.44        95.51        1.06x
BenchmarkMVCCScan1Version10Rows       280.04       376.60       1.34x
BenchmarkMVCCScan1Version100Rows      370.69       471.74       1.27x
BenchmarkMVCCScan1Version1000Rows     464.35       558.09       1.20x
BenchmarkMVCCGet1Version              273.83       324.77       1.19x

bdarnell commented on storage/engine/engine.go:
```diff
@@ -188,6 +188,13 @@ func PutProto(engine Engine, key proto.EncodedKey, msg gogoproto.Message) (keyBy
 // key was not found. On success, returns the length in bytes of the
 // key and the value.
 func GetProto(engine Engine, key proto.EncodedKey, msg gogoproto.Message) (ok bool, keyBytes, valBytes int64, err error) {
+	type protoGetter interface {
```
Any interfaces used should generally be public and documented; it can be surprising when you pass an Engine to a function and a method that is not a part of the Engine interface is called. Or we could just add GetProto to the Engine interface to ensure that Batch and Snapshot implement it and pass it along. (there's no good way for wrappers to implement protoGetter iff their wrapped object does).

Instead of a new method on Engine, we may want to change the interface of Engine.Get to take an object to receive the bytes (similar to json.Unmarshaler) instead of returning the bytes. There aren't that many references to Engine.Get so it shouldn't be too hard to change.

petermattis replied:
I like the `Unmarshaler` idea. I'll play around with that tomorrow.

petermattis replied:
I decided to just add `GetProto` to `Engine` and `ValueProto` to `Iterator`. Not as general as the `Unmarshaler` idea, but simpler.

bdarnell commented on storage/engine/rocksdb.go:
```diff
@@ -166,6 +167,40 @@ func (r *RocksDB) getInternal(key proto.EncodedKey, snapshotHandle *C.DBSnapshot
+		// Make a byte slice that is backed by result.data. This slice
+		// cannot live past the lifetime of this method, but we're only
+		// using it to unmarshal the proto.
+		// data := C.GoBytes(unsafe.Pointer(result.data), result.len)
```
Either explain why this commented-out line is here or remove it.

The assumption that Unmarshal always copies any bytes it uses from the input data scares me. It should generally be true since Unmarshal doesn't want to keep the input data from getting GC'd, but this greatly increases the stakes for any code paths that simply slice bytes out of the input without copying. I'm not sure the performance improvements here are worth the loss of memory safety.

petermattis replied:
The commented out line was for testing. I'll delete. Unmarshal definitely always copies any bytes it uses from the input data. I can add a test to this effect. It would be a fairly horrible API if Unmarshal did not copy input data it wanted to keep as doing so would allow the caller to accidentally reuse the input data and end up changing the decoded data due to the sharing.

bdarnell replied:
Good point. I guess using a raw pointer behind the slice doesn't really increase the risk of reusing the backing array compared to what can already happen with pure Go slices, so I withdraw my concern here.

bdarnell commented on storage/engine/rocksdb.go:
```diff
@@ -166,6 +167,40 @@ func (r *RocksDB) getInternal(key proto.EncodedKey, snapshotHandle *C.DBSnapshot
+		// cannot live past the lifetime of this method, but we're only
+		// using it to unmarshal the proto.
+		// data := C.GoBytes(unsafe.Pointer(result.data), result.len)
+		data := (*[0x7fffffff]byte)(unsafe.Pointer(result.data))[:result.len:result.len]
```
Can you pull this out into a function? The 0x7fffffff functions as a 2GB limit on the size of data we can support, right? Is there any way we can avoid a magic constant here? Or make it `2**63-1` instead of `2**31-1`?

petermattis replied:
I'll pull it out into a separate function. I got this from a stackoverflow post or from go-nuts and there was a rationale behind the 2GB limit that I can't remember.

Do we have other limits on the size of data?

bdarnell replied:
I don't think we currently have any hard limits on data size, although as a practical matter we'll want to set some limits far below 2GB. I'm not too concerned about hitting the 2GB limit (especially for protobufs; anything that big will probably be a blob instead of a protobuf), I'd just rather not have arbitrary numbers in the code that represent "more than we'll ever need" if we can avoid it.

petermattis replied:
I moved this code into `cSliceToUnsafeGoBytes`, explained where the constant `0x7fffffff` is coming from and fall back to using `cSliceToGoBytes` when the unsafe approach is not possible (arrays cannot be larger than 2^31-1 entries).

petermattis committed:
Remove the use of private interfaces.

Add Engine.GetProto and Iterator.ValueProto functions to the Engine and
Iterator interfaces. Provide implementations of these functions for the
existing engines and iterators.

petermattis commented:
Not quite ready for another look. I have to track down the stuff about a 2GB data limit and what the appropriate constant is to use there.

petermattis committed:
Add cSliceToUnsafeGoBytes.

petermattis commented:
This is ready for another look.

bdarnell commented:
LGTM

mrtracy commented on storage/engine/mvcc.go:
```diff
@@ -432,53 +444,71 @@ func MVCCPutProto(engine Engine, ms *MVCCStats, key proto.Key, timestamp proto.T
+	metaKey := mvccEncodeKey(buf.key[0:0], key)
+	ok, _, _, err := engine.GetProto(metaKey, &buf.meta)
+	if err != nil || !ok {
+		getBufferPool.Put(buf)
```
Any reason not to defer this?

petermattis replied:
None. Switched to using defer. The perf numbers are unchanged.

petermattis committed:
Use defer for re-adding the buffer to the getBufferPool.

mrtracy commented on storage/engine/mvcc.go:
```diff
@@ -432,53 +444,71 @@ func MVCCPutProto(engine Engine, ms *MVCCStats, key proto.Key, timestamp proto.T
 // timestamp of the earlier version, a boolean indicating whether a
 // version value or metadata was found, and error, if applicable.
-type getEarlierFunc func(engine Engine, start, end proto.EncodedKey) (proto.RawKeyValue, error)
+type getValueFunc func(engine Engine, start, end proto.EncodedKey,
```
getEarlierFunc seemed like a clearer name to me.

petermattis replied:
I was confused by the previous code that `getEarlierFunc` was being used to retrieve both the metadata and the values. The `getValueFunc` is only used to retrieve values. This seems cleaner to me.

mrtracy commented on storage/engine/mvcc.go:
```diff
@@ -872,31 +906,48 @@ func MVCCScan(engine Engine, key, endKey proto.Key, max int64, timestamp proto.T
+	defer getBufferPool.Put(buf)
+
+	encEndKey := mvccEncodeKey(buf.key[0:0], endKey)
+	keyBuf := buf.key[len(encEndKey):len(encEndKey)]
```
It looks like this will result in an index error if endKey is longer than 1024 bytes.

petermattis replied:
Good catch. Fixed.

petermattis committed:
Fix bug in handling long keys in MVCCScan.
